### PR TITLE
Check if Rust lib needs rebuilding.

### PR DESCRIPTION
**Pull request by erikjohnston**
This protects against the common mistake of failing to remember to rebuild Rust code after making changes.

### CODE REVIEW

File: rust/build.rs [hunk 1]
```diff
@@ -0,0 +1,44 @@
+use std::path::PathBuf;
+
+use blake2::{Blake2b512, Digest};
+
+fn main() -> Result<(), std::io::Error> {
```
Comment (erikjohnston) [inline rust/build.rs:5]: Oh, I should add a comment as to why we're doing this.

File: rust/build.rs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    paths.sort();
+
+    for path in paths {
+        let bytes = std::fs::read(path)?;
+        hasher.update(bytes);
+    }
+
+    let hex_digest = hex::encode(hasher.finalize());
+    println!("cargo:warning={hex_digest}");
```
Comment (erikjohnston) [inline rust/build.rs:40]:
```suggestion
```

File: rust/build.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+        for entry in entries {
+            if entry.is_dir() {
+                new_dirs.push(entry);
+            } else {
+                paths.push(entry.to_str().expect("valid rust paths").to_string());
+            }
+        }
+
+        dirs.append(&mut new_dirs);
```
Comment (squahtx) [inline rust/build.rs:33]: As someone who has written fewer than 10 lines of Rust, why can't we append to `dirs` directly inside the loop?

**Review by squahtx — APPROVED** [review]
Looks reasonable!

File: rust/build.rs [hunk 4]
```diff
@@ -0,0 +1,49 @@
+//! This build script is calculates the hash of all files in the `src/`
+//! directory and adds it as an environment variable during build time.
```
Comment (clokep) [inline rust/build.rs:2]:
```suggestion
//! This build script calculates the hash of all files in the `src/`
//! directory and adds it as an environment variable during build time.
```

File: rust/build.rs (hunk 3 above)
Comment (erikjohnston) [reply rust/build.rs:33]: Woops, this was a holdover from a previous iteration of the code. Will move it inside!

**Pull request merged**